### PR TITLE
Fix for issue 5355 compiler.get_supported_argument reports success on unsupported arguments

### PR DESCRIPTION
**Pull request by kayakMike**
updated the ClikeCompiler.has_argument method to parse the stderr of the compiler check
   to ensure the compile argument is truely supported by the cpp compiler.

### CODE REVIEW

File: mesonbuild/compilers/mixins/clike.py [hunk 1]
```diff
@@ -1069,7 +1069,12 @@ def linker_to_compiler_args(self, args):
         return args
 
     def has_arguments(self, args, env, code, mode):
-        return self.compiles(code, env, extra_args=args, mode=mode)
+        with self._build_wrapper(code, env, args, None, mode, disable_cache=False, want_output=True) as p:
+            stdout, stderr = p.communicate()
+            result = p.returncode == 0
+            if self.language == "cpp" and "is valid for C/ObjC" in stderr.decode():
+                result = False
+            return result, p.cached
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/mixins/clike.py]: __[Flake8]__ [W291] trailing whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5866#issue-4498794))

<sub>posted by Sider</sub>

Comment (kayakMike) [reply mesonbuild/compilers/mixins/clike.py]: ah...  right, first time mistake, sorry.

**Comment by dcbaker** [comment]
I think this belongs in `mesonbuild/compilers/mixins/gnu.py` in the `GnuCompiler` class. I wonder if we could get the clang and gcc developers to just give as an argument like `-Werror=bad-flag` or something so that we wouldn't need to do this...

**Comment by kayakMike** [comment]
@dcbaker agreed as its a work-around for gnu specific message on stderr, so it does seem more appropriate to go in the gnu.py mixin.   I'll submit again shortly.

File: mesonbuild/compilers/mixins/gnu.py [hunk 2]
```diff
@@ -337,3 +337,19 @@ def get_pch_suffix(self) -> str:
 
     def openmp_flags(self) -> typing.List[str]:
         return ['-fopenmp']
+
+    def has_arguments(self, args, env, code, mode):
+        # For some compiler command line arguments, the GNU compilers will 
+        # emit a warning on stderr indicating that an option is valid for a
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/mixins/gnu.py]: __[Flake8]__ [W291] trailing whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5866#issue-4519091))

<sub>posted by Sider</sub>

File: mesonbuild/compilers/mixins/gnu.py [hunk 3]
```diff
@@ -337,3 +337,19 @@ def get_pch_suffix(self) -> str:
 
     def openmp_flags(self) -> typing.List[str]:
         return ['-fopenmp']
+
+    def has_arguments(self, args, env, code, mode):
+        # For some compiler command line arguments, the GNU compilers will
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/mixins/gnu.py]: __[Flake8]__ [W291] trailing whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5866#issue-4519090))

<sub>posted by Sider</sub>

File: mesonbuild/compilers/mixins/gnu.py [hunk 4]
```diff
@@ -337,3 +337,19 @@ def get_pch_suffix(self) -> str:
...
+        # emit a warning on stderr indicating that an option is valid for a
+        # another language, but still complete with exit_success
+        with self._build_wrapper(code, env, args, None, mode, disable_cache=False, want_output=True) as p:
+            result = p.returncode == 0
+            try:
+                stdout, stderr = p.communicate()
+                if self.language == "cpp" and "is valid for C/ObjC" in stderr.decode():
```
Comment (dcbaker) [inline mesonbuild/compilers/mixins/gnu.py]: this should probably be `if self.language in {'cpp', 'objcpp'}`

File: mesonbuild/compilers/mixins/gnu.py [hunk 5]
```diff
@@ -337,3 +337,19 @@ def get_pch_suffix(self) -> str:
...
+                    result = False
+                if self.language == "c" and "is valid for C++/ObjC++" in stderr.decode():
```
Comment (dcbaker) [inline mesonbuild/compilers/mixins/gnu.py]: and this should probably be `if self.language in {'c', 'objc'}`

Comment (kayakMike) [reply mesonbuild/compilers/mixins/gnu.py]: Thanks, appreciate the suggestion, its better this way.  Rebased and force pushed.

File: mesonbuild/compilers/mixins/gnu.py (hunk 4 above)
Comment (kayakMike) [reply mesonbuild/compilers/mixins/gnu.py]: Indeed.  Appreciate all the extra help.  I also found a better way to get the error message so I could drop the try - except clause.  I think this also could be a change brought in on my latest rebase.

**Comment by kayakMike** [comment]
@dcbaker  Did I miss any change that you might want to see?

**Comment by jpakkane** [comment]
@dcbaker is this ok by you know?

**Comment by kayakMike** [comment]
Every time I rebase master on this branch to keep it current; it triggers the CI/CD system.  I don't know if this incurs some cost to the project as testing VMs may not be free resources.

**Comment by jpakkane** [comment]
We only use the free tier, so no. But CI gets clogged up every now and then so it makes sense to minimize the number of rebases.

**Pull request merged**